[JIT] Support for NamedTuple

jamesr66a opened the pull request:
Resolves https://github.com/pytorch/lockdown/issues/18

This implements NamedTuple by taking advantage of the existing `names` field in `TupleType`. 

TODO: This currently doesn't retain the NamedTuple-ness through serialization. Discussed with @suo offline, we can probably make a way to define an anonymous NamedTuple in script (e.g. `NamedTuple('Foo', [('a', int), ('b', float), ('c', List[float])])` and serialize that
TODO: implement support for calling the constructor with kwargs

jamesr66a commented on test/test_jit_py3.py:
```diff
@@ -69,3 +70,42 @@ def func(x):
+
+        @torch.jit.script
+        def foo(x) -> float:
+            fv = FeatureVector(3.0, [3.0], 3.0)  # noqa
```
flake8 was freaking out here in my pre-commit hook for some reason

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -620,6 +628,30 @@ void initJitScriptBindings(PyObject* module) {
+         const std::map<std::string, TypePtr> annotations,
+         ResolutionCallback rcb) {
+        auto cu = std::make_shared<CompilationUnit>();
+        auto classType =
```
this code is dead

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -620,6 +628,30 @@ void initJitScriptBindings(PyObject* module) {
+  m.def(
+      "_jit_script_named_tuple_compile",
+      [](const std::string& qualifiedName,
+         const std::vector<std::string>& fields,
```
we should be passing a `TreeView` here, similar to how the other compile functions work. I think the python side should be responsible for lowering something like `namedtuple("foo", ["bar", str])` into the right `TreeView` (probably a classdef?).

This will ensure that when we eventually merge the two concepts they aren't too out of sync

suo commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -192,9 +193,21 @@ struct GenericDict;
 
 struct CAFFE2_API Tuple : public List<IValue> {
   using List<IValue>::List;
+
```
btw, in your serialization PR you'll need to fix up the pickler as well—currently the `type` member will just get ignore and it'll get pickled as a tuple, which isn't what we want.

suo reviewed (changes requested):
This duplication is sad, but I can't really see a better way short of merging the tuple and class concepts together. Can you add a lockdown P1 to explore whether that's possible? I think it should work if we implement a scalar replacement of aggregates pass.

Otherwise, left some comments inline

suo commented on torch/jit/__init__.py:
```diff
@@ -1071,6 +1072,17 @@ def script(obj, optimize=True, _frames_up=0, _rcb=None):
+            else:
+                annotations = OrderedDict()
+            _jit_script_named_tuple_compile(qualified_name, fields, annotations, _rcb)
+            _add_script_class(obj, qualified_name)
```
I realized we don't actually need any of these name -> type mappings at all (we don't need the one in CompilationUnit either). Since we're structurally typing NamedTuples, all we need is a "Python namedtuple -> JIT TupleType" translation. When the compiler encounters a type `Foo`, it should just look up `Foo` in Python environment; if it's a named tuple, generate the corresponding tuple type, and use that.

suo commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -104,6 +110,17 @@ struct PythonResolver : public Resolver {
       return nullptr;
     }
 
+    if (isTupleClass(obj)) {
```
should this not be `isNamedTupleClass`?

suo commented on torch/jit/__init__.py:
```diff
@@ -2013,6 +2013,21 @@ def _get_script_class(name):
+    return fields, annotations
+
+def _create_named_tuple(t, names):
+    TupleType = collections.namedtuple('NamedTuple', names)
```
We should still be extracting the unqualified class name in `_get_named_tuple_properties`, putting it into TupleType and restoring it correctly here.

suo reviewed (commented):
Looks good! Much simpler than before. Only one substantive comment

suo reviewed (approved):
lgtm. I only glanced the tupleptr reverts, I assume they are just straight reverts. In the future, please stack your PRs so they are easier to review 💯

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -264,6 +264,36 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
+    throw ErrorReport(loc) << "Constructor expected " << nargs << "arguments "
+                           << "but got " << inputs.size();
+  }
+  AT_CHECK(attributes.size() == 0);
```
`MyNamedTuple(a=3) ` is going to error without line information.
This patch: make this an error report. In the future, we should be supporting keyword construction of tuples.

zdevito commented on torch/csrc/jit/script/sugared_value.h:
```diff
@@ -215,6 +215,23 @@ struct TORCH_API ClassValue : public SugaredValue {
   ClassTypePtr type_;
 };
 
+struct TORCH_API NamedTupleValue : public SugaredValue {
```
nit: NamedTupleConstructor

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -264,6 +264,36 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
+                                input_vals, type_->names(), type_->unqualName())
+                               ->setSourceRange(loc))
+                  ->output();
+  self->setType(type_);
```
Uhhh, there is absolutely no type checking here!

zdevito replied:
I suggest that a FunctionSchema is created that is the right constructor list for the named tuple, and then matchSchema is used. That fixes the keyword arguments, and does the type checking correctly.

zdevito commented on test/test_jit_py3.py:
```diff
@@ -69,3 +71,82 @@ def func(x):
+
+        mnt = MyCoolNamedTuple(42, 420.0, [666])
+        self.assertEqual(foo(mnt), mnt)
+
```
can we add a test where the wrong arguments are passed to the constructor?

zdevito reviewed (commented):
The structure looks good, but I don't there is any type checking being done on tuple construction.
I'd also like to see what happens on unification e.g. a named and unnamed tuple on either side of an if-statement.

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -264,6 +264,59 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
+      /*arguments=*/arguments,
+      /*returns=*/{});
+
+  matchSchema(schema, loc, g, inputs, attributes);
```
Look at other uses of matchSchema. This needs to take the _output_ of match schema and pass it into tuple construct. Otherwise, this introduces type errors (implicit conversions were run and discarded). Also this would trivially handle keyword arguments. Take a look at the other places where matchSchema is used.

zdevito reviewed (dismissed):
Schema checking code is buggy. It is an easy fix but this isn't correct yet.

zdevito commented on torch/csrc/jit/script/sugared_value.cpp:
```diff
@@ -264,6 +264,59 @@ std::shared_ptr<SugaredValue> ClassValue::attr(
+
+  auto& g = *m.graph();
+
+  // Check that the types of the inputs passed into the NamedTuple constructor
```
The thing that generates the FunctionSchema should go in its own function, probably a method of TupleType. It is not great that it is recalculating the FunctionSchema everytime we try to construct a member.

zdevito reviewed (commented):
Schema checking code is buggy. It is an easy fix but this isn't correct yet.

Commit:
Support for NamedTuple (#21428)

Summary:
Resolves https://github.com/pytorch/lockdown/issues/18

This implements NamedTuple by taking advantage of the existing `names` field in `TupleType`.

TODO: This currently doesn't retain the NamedTuple-ness through serialization. Discussed with suo offline, we can probably make a way to define an anonymous NamedTuple in script (e.g. `NamedTuple('Foo', [('a', int), ('b', float), ('c', List[float])])` and serialize that
TODO: implement support for calling the constructor with kwargs
Pull Request resolved: https://github.com/pytorch/pytorch/pull/21428

Differential Revision: D15741564

fbshipit-source-id: b981d3c0f058afec5d6a7500d989bb10ac898278
